### PR TITLE
PR for issue/35

### PR DESCRIPTION
**Pull request by FRosner**
Rebased version of https://github.com/FRosner/drunken-data-quality/pull/54

### CODE REVIEW

File: src/main/scala/de/frosner/ddq/core/Check.scala [hunk 1]
```diff
@@ -563,13 +563,16 @@ object Check {
...
+      val join = renamedDfDistinct.join(renamedRef, renamedColumns.map{
         case (baseColumn, refColumn) => new Column(baseColumn) === new Column(refColumn)
       }.reduce(_ && _))
-      val matchingRows = join.count
+      val matchingRows = join.distinct.count
+      val mergeRate = matchingRows.toFloat/distinctBefore
       val columnsString = columns.map{ case (baseCol, refCol) => baseCol + "->" + refCol }.mkString(", ")
       if (matchingRows > 0)
-        ConstraintSuccess(s"${if(columns.length == 1) "Column" else "Columns"} $columnsString can be used for joining (${ if(matchingRows == 1) "one distinct row" else s"$matchingRows distinct rows"} match)")
+        ConstraintSuccess(f"""${if(columns.length == 1) "Column" else "Columns"} $columnsString can be used for joining (number of distinct rows in base table: $distinctBefore, number of distinct rows after joining: $matchingRows, merge rate: $mergeRate%.2f)""")
```
Comment (FRosner) [inline src/main/scala/de/frosner/ddq/core/Check.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1991422054/issues/source?bid=3033737&fileBranchId=3048638#l575)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 2]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
     val ref = makeIntegersDf(List(1, 3, 100), List(1, 500, 100))
     val check = Check(base).isJoinableWith(ref, "column1" -> "column1", "column3" -> "column2")
     val constraint = check.constraints.head
-    val result = ConstraintSuccess("Columns column1->column1, column3->column2 can be used for joining (one distinct row match)")
+    val result = ConstraintSuccess("Columns column1->column1, column3->column2 can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, merge rate: 0.50)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1991422063/issues/source?bid=3033737&fileBranchId=3048638#l492)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 3]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
     check.run().constraintResults shouldBe Map(constraint -> result)
   }
 
+  it should "compute the merge rate in an commutative way" in {
+    val base = makeIntegerDf(List(1, 1, 1, 1, 1, 1, 1, 1, 1, 2))
+    val ref = makeIntegerDf(List(1))
+
+    val check1 = Check(base).isJoinableWith(ref, "column" -> "column")
+    val constraint1 = check1.constraints.head
+    val result1 = ConstraintSuccess("Column column->column can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, merge rate: 0.50)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1991422063/issues/source?bid=3033737&fileBranchId=3048638#l502)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 4]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+    check1.run().constraintResults shouldBe Map(constraint1 -> result1)
+
+    val check2 = Check(ref).isJoinableWith(base, "column" -> "column")
+    val constraint2 = check2.constraints.head
+    val result2 = ConstraintSuccess("Column column->column can be used for joining (number of distinct rows in base table: 1, number of distinct rows after joining: 1, merge rate: 0.50)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1991422063/issues/source?bid=3033737&fileBranchId=3048638#l507)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 5]
```diff
@@ -480,7 +480,7 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
     val ref = makeIntegersDf(List(1, 2, 100), List(1, 5, 100))
     val check = Check(base).isJoinableWith(ref, "column1" -> "column1", "column2" -> "column2")
     val constraint = check.constraints.head
-    val result = ConstraintSuccess("Columns column1->column1, column2->column2 can be used for joining (one distinct row match)")
+    val result = ConstraintSuccess("Columns column1->column1, column2->column2 can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, merge rate: 0.50)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1991422063/issues/source?bid=3033737&fileBranchId=3048638#l483)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 6]
```diff
@@ -471,7 +471,7 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
     val ref = makeIntegerDf(List(1, 2, 5))
     val check = Check(base).isJoinableWith(ref, "column" -> "column")
     val constraint = check.constraints.head
-    val result = ConstraintSuccess("Column column->column can be used for joining (2 distinct rows match)")
+    val result = ConstraintSuccess("Column column->column can be used for joining (number of distinct rows in base table: 3, number of distinct rows after joining: 2, merge rate: 0.67)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1991422063/issues/source?bid=3033737&fileBranchId=3048638#l474)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 7]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+    val result = ConstraintSuccess("Columns column1->column1, column3->column2 can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, merge rate: 0.50)")
     check.run().constraintResults shouldBe Map(constraint -> result)
   }
 
+  it should "compute the merge rate in an commutative way" in {
+    val base = makeIntegerDf(List(1, 1, 1, 1, 1, 1, 1, 1, 1, 2))
+    val ref = makeIntegerDf(List(1))
+
+    val check1 = Check(base).isJoinableWith(ref, "column" -> "column")
+    val constraint1 = check1.constraints.head
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1991422063/issues/source?bid=3033737&fileBranchId=3048638#l501)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 8]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+    val result1 = ConstraintSuccess("Column column->column can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, merge rate: 0.50)")
+    check1.run().constraintResults shouldBe Map(constraint1 -> result1)
+
+    val check2 = Check(ref).isJoinableWith(base, "column" -> "column")
+    val constraint2 = check2.constraints.head
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1991422063/issues/source?bid=3033737&fileBranchId=3048638#l506)

**Comment by ghost** [comment]
@FRosner Ok, the merge rate basically tells me the percentage of "survived" (distinct) primary keys after the join with some other table. I just googled again for the term 'merge rate' with respect to joins/SQL but was not successful in finding an official definition. So I think we can just leave it like this and document it properly.

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 9]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+    val result = ConstraintSuccess("Columns column1->column1, column3->column2 can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, unmatched keys in base table: 50%)")
     check.run().constraintResults shouldBe Map(constraint -> result)
   }
 
+  it should "compute the unmatched keys in base table correctly" in {
+    val base = makeIntegerDf(List(1, 1, 1, 1, 1, 1, 1, 1, 1, 2))
+    val ref = makeIntegerDf(List(1))
+
+    val check1 = Check(base).isJoinableWith(ref, "column" -> "column")
+    val constraint1 = check1.constraints.head
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1994540670/issues/source?bid=3033737&fileBranchId=3048638#l501)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 10]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+    val result1 = ConstraintSuccess("Column column->column can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, unmatched keys in base table: 50%)")
+    check1.run().constraintResults shouldBe Map(constraint1 -> result1)
+
+    val check2 = Check(ref).isJoinableWith(base, "column" -> "column")
+    val constraint2 = check2.constraints.head
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1994540670/issues/source?bid=3033737&fileBranchId=3048638#l506)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 11]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
     val ref = makeIntegersDf(List(1, 3, 100), List(1, 500, 100))
     val check = Check(base).isJoinableWith(ref, "column1" -> "column1", "column3" -> "column2")
     val constraint = check.constraints.head
-    val result = ConstraintSuccess("Columns column1->column1, column3->column2 can be used for joining (one distinct row match)")
+    val result = ConstraintSuccess("Columns column1->column1, column3->column2 can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, unmatched keys in base table: 50%)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1994540670/issues/source?bid=3033737&fileBranchId=3048638#l492)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 12]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+    val ref = makeIntegerDf(List(1))
+
+    val check1 = Check(base).isJoinableWith(ref, "column" -> "column")
+    val constraint1 = check1.constraints.head
+    val result1 = ConstraintSuccess("Column column->column can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, unmatched keys in base table: 50%)")
+    check1.run().constraintResults shouldBe Map(constraint1 -> result1)
+
+    val check2 = Check(ref).isJoinableWith(base, "column" -> "column")
+    val constraint2 = check2.constraints.head
+    val result2 = ConstraintSuccess("Column column->column can be used for joining (number of distinct rows in base table: 1, number of distinct rows after joining: 1, unmatched keys in base table: 100%)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1994540670/issues/source?bid=3033737&fileBranchId=3048638#l507)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 13]
```diff
@@ -480,7 +480,7 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
     val ref = makeIntegersDf(List(1, 2, 100), List(1, 5, 100))
     val check = Check(base).isJoinableWith(ref, "column1" -> "column1", "column2" -> "column2")
     val constraint = check.constraints.head
-    val result = ConstraintSuccess("Columns column1->column1, column2->column2 can be used for joining (one distinct row match)")
+    val result = ConstraintSuccess("Columns column1->column1, column2->column2 can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, unmatched keys in base table: 50%)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1994540670/issues/source?bid=3033737&fileBranchId=3048638#l483)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 14]
```diff
@@ -471,7 +471,7 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
     val ref = makeIntegerDf(List(1, 2, 5))
     val check = Check(base).isJoinableWith(ref, "column" -> "column")
     val constraint = check.constraints.head
-    val result = ConstraintSuccess("Column column->column can be used for joining (2 distinct rows match)")
+    val result = ConstraintSuccess("Column column->column can be used for joining (number of distinct rows in base table: 3, number of distinct rows after joining: 2, unmatched keys in base table: 67%)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1994540670/issues/source?bid=3033737&fileBranchId=3048638#l474)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 15]
```diff
@@ -489,10 +489,25 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
     check.run().constraintResults shouldBe Map(constraint -> result)
   }
 
+  it should "compute the unmatched keys in base table correctly" in {
+    val base = makeIntegerDf(List(1, 1, 1, 1, 1, 1, 1, 1, 1, 2))
+    val ref = makeIntegerDf(List(1))
+
+    val check1 = Check(base).isJoinableWith(ref, "column" -> "column")
+    val constraint1 = check1.constraints.head
+    val result1 = ConstraintSuccess("Column column->column can be used for joining (number of distinct rows in base table: 2, number of distinct rows after joining: 1, unmatched keys in base table: 50%)")
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1994540670/issues/source?bid=3033737&fileBranchId=3048638#l502)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 16]
```diff
@@ -489,10 +493,31 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+      "join columns cardinality after joining: 1 (50%)")
     check.run().constraintResults shouldBe Map(constraint -> result)
   }
 
+  it should "compute the matched keys in base table correctly" in {
+    val base = makeIntegerDf(List(1, 1, 1, 1, 1, 1, 1, 1, 1, 2))
+    val ref = makeIntegerDf(List(1))
+
+    val check1 = Check(base).isJoinableWith(ref, "column" -> "column")
+    val constraint1 = check1.constraints.head
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1996207281/issues/source?bid=3033737&fileBranchId=3048638#l507)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 17]
```diff
@@ -489,10 +493,31 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+    val result1 = ConstraintSuccess("Column column->column can be used for joining (" +
+      "join columns cardinality in base table: 2, " +
+      "join columns cardinality after joining: 1 (50%)")
+    check1.run().constraintResults shouldBe Map(constraint1 -> result1)
+
+    val check2 = Check(ref).isJoinableWith(base, "column" -> "column")
+    val constraint2 = check2.constraints.head
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1996207281/issues/source?bid=3033737&fileBranchId=3048638#l514)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 18]
```diff
@@ -489,10 +493,31 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+      "join columns cardinality after joining: 1 (50%)")
     check.run().constraintResults shouldBe Map(constraint -> result)
   }
 
+  it should "compute the matched keys in a non-commutative way" in {
+    val base = makeIntegerDf(List(1, 1, 1, 1, 1, 1, 1, 1, 1, 2))
+    val ref = makeIntegerDf(List(1))
+
+    val check1 = Check(base).isJoinableWith(ref, "column" -> "column")
+    val constraint1 = check1.constraints.head
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1998406483/issues/source?bid=3033737&fileBranchId=3048638#l507)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala [hunk 19]
```diff
@@ -489,10 +493,31 @@ class CheckTest extends FlatSpec with Matchers with BeforeAndAfterEach with Befo
...
+    val result1 = ConstraintSuccess("Column column->column can be used for joining (" +
+      "join columns cardinality in base table: 2, " +
+      "join columns cardinality after joining: 1 (50%)")
+    check1.run().constraintResults shouldBe Map(constraint1 -> result1)
+
+    val check2 = Check(ref).isJoinableWith(base, "column" -> "column")
+    val constraint2 = check2.constraints.head
```
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1998406483/issues/source?bid=3033737&fileBranchId=3048638#l514)

**Comment by FRosner** [comment]
Can you please review the PR one more time @bkomboz @zchen?

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala (hunk 18 above)
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1998678313/issues/source?bid=3033737&fileBranchId=3048638#l507)

File: src/test/scala/de/frosner/ddq/core/CheckTest.scala (hunk 19 above)
Comment (FRosner) [inline src/test/scala/de/frosner/ddq/core/CheckTest.scala]: Codacy Issue found: [Usage of head on collections.](https://www.codacy.com/app/frank_7/drunken-data-quality/file/1998678313/issues/source?bid=3033737&fileBranchId=3048638#l514)

**Comment by ghost** [comment]
@FRosner Overall it looks good to me. Minor detail: Should variable `unmatchedKeysPercentage` not  better be named `matchedKeysPercentage`? Because the following test:

```
"A joinable check" should "succeed if a join on the given column yields at least one row" in {
val base = makeIntegerDf(List(1, 1, 1, 2, 2, 3))
val ref = makeIntegerDf(List(1, 2, 5))
val check = Check(base).isJoinableWith(ref, "column" -> "column")
val constraint = check.constraints.head
val result = ConstraintSuccess("Column column->column can be used for joining (" +
  "join columns cardinality in base table: 3, " +
  "join columns cardinality after joining: 2 (67%)")
check.run().constraintResults shouldBe Map(constraint -> result)
}
```

gives 67% (percentage of matches keys) and not 33% (percentage of unmatched keys).

**Comment by FRosner** [comment]
You're actually right. Thanks!

**Pull request merged**